src/event_manager/mod.rs

```diff
@@ -0,0 +1,58 @@
+mod control_ops;
+mod endpoint;
+mod epoll_context;
+mod events;
+mod manager;
+mod subscribers;
+
+use std::io;
+use std::result;
+
+pub use control_ops::ControlOps;
+pub use endpoint::RemoteEndpoint;
+pub use events::Events;
+pub use manager::EventManager;
+
+/// Error conditions that may appear during `EventManager` related operations.
+#[derive(Debug)]
+pub enum Error {
+    ChannelSend,
+    ChannelRecv,
+    Epoll(io::Error),
+    EventFd(io::Error),
+    // TODO: should we allow fds to be registered multiple times?
+    FdAlreadyRegistered,
+    InvalidId,
+}
+
+/// Generic result type that may return `EventManager` errors.
+pub type Result<T> = result::Result<T, Error>;
+
+/// Opaque object that uniquely represents a subscriber registered with an `EventManager`.
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub struct SubscriberId(u128);
+
+/// The `EventSubscriber` trait allows the interaction between an `EventManager` and different
+/// event subscribers.
+pub trait EventSubscriber {
+    /// Respond to events and potentially alter the interest set of the subscriber.
+    ///
+    /// Called by the `EventManager` whenever an event associated with the subscriber is triggered.
+    fn process(&mut self, events: Events, ops: &mut ControlOps);
+
+    /// Register the events initially associated with the subscriber.
+    ///
+    /// Called by the `EventManager` after a subscriber is registered.
+    fn init(&self, ops: &mut ControlOps);
```

Comment:
> how about taking a" &mut self"?
Hmmm, I am not sure you need to mutate self. Do you? If that's the case, we can change it to mut.

`ControlOps` was intentionally added as an abstraction here to fix point 2 in the TODO list: https://github.com/rust-vmm/vmm-epoll/pull/2#issuecomment-586204636.

The idea behind `ControlOps` was to hide the fact that the event manager works with `epoll` and also limit the operations that you can do from the subscriber. This way we make sure that only valid operations are called: `add`, `remove`, `modify`.

Reply:
> > how about taking a" &mut self"?
> 
> Hmmm, I am not sure you need to mutate self. Do you? If that's the case, we can change it to mut.
> 
> `ControlOps` was intentionally added as an abstraction here to fix point 2 in the TODO list: [#2 (comment)](https://github.com/rust-vmm/vmm-epoll/pull/2#issuecomment-586204636).
> 
> The idea behind `ControlOps` was to hide the fact that the event manager works with `epoll` and also limit the operations that you can do from the subscriber. This way we make sure that only valid operations are called: `add`, `remove`, `modify`.

We could achieve the goal of [#2](https://github.com/rust-vmm/vmm-epoll/pull/2#issuecomment-586204636) by
* add argument "subscriber_id" to EventSubscriber::init()
* make EventSubscriber::init() takes &mut self
* export EpollContext directly
* move add/modify/remove from ControlOps to EpollContext and make them pub.
By this way, it gets simpler and achieves a minor performance improve.

Reply:
I don't like the idea of exporting `EpollContext` with its current form and name because it `Context` doesn't really mean anything and `Epoll` is strongly suggesting that everything is based on `epoll` :)). I will try to find an abstraction that makes more sense at the interface level, while keeping in mine your feedback. To Alex's comment https://github.com/rust-vmm/vmm-epoll/pull/2#discussion_r402621389, what do you think about renaming this crate to `event-manager` instead? This way we could leave it open for other implementations as well. The interface could be extended to work on Windows as well or support other "backends". WDYT?

Reply:
event-manager is OK to me, to be more agnostic.

Reply:
Cool. Event Manager seems to be available on crates.io as well.

Reply:
@jiangliu with this approach, you would need to check the validity of the <SubscriberId, Events> pair when calling process. I didn't do any benchmarks, but I assume that the minor performance improvement will not be visible anymore because of the extra checks that are needed.

Reply:
For most epoll events, the handler will do read/write operations instead of EventOperations::{add, remove, modify}.

Reply:
Hmm, do you mean that during process you are mostly just processing events and not using the epoll reference to alter the events?